doc/dev/crimson/pipelines.rst

```diff
@@ -0,0 +1,258 @@
+client requests?
+
+Let's take a look at a use case: a certain client sends 10 consecutive requests targeting different objects.
+
+Since these requests are targeting different objects, The status of the objects would be eventually consistent
+no matter in which order the write requests are executed, even if these requests are writes. It sounds like a
+nice scatter/gather I/O at a higher level. But this does not really happen in real world. Because, most librados
+applications, RBD for instance, do not work in terms of a single huge RADOS object. They slice a large object
+into smaller fixed-size chunks based on offset, and store these chunks as bounded size RADOS objects. An RBD
+image is always mapped to multiple RADOS objects, which are in turn placed in different PGs hosted by differnt
+OSDs. So, 10 requests on an RBD device won't be pipelined in general. It's more unlikely than getting 10 heads
+in row when tossing coins.
+
+
+More than 1 Visitor in a Room
+-----------------------------
+
+Some stages in the pipeline above do not really "process" requests. They are more like gatekeepers.
+For example, if a client request is waiting for the OSD to be updated with an updated osdmap, its
+next request will also very likely to be waiting for the same map if it is allowed to enter that stage.
+So, in this case, allowing multiple request from the same client to wait for the same gatekeepers in
+the same stage does not help.
+
+But the last stage in the pipeline does process requests. The "process" stage calls into the PG
+backend for performing I/O. It even recovers the object if the object to be accessed is not available
+or corrupted. "Process" stage is a per-PG pharse. Namely, each PG has only a single "room" for
+processing requests. So, a PG is not able to handle multiple operations at a time due to this design.
+
+- for multiple write requests targeting the same object.
+
+  It really doesn't make sense to execute multiple writes on the same object at once. Because we already
+  do the only sensible version by chopping up an RBD image into multiple 4MB chunks. We are not likely
+  to benefit from doing so.
+- for multiple write requests targeting different object. These requests are sent by different clients.
+
+  We will have to ensure the ACID of the object store with a more sophisticated design.
+- for multiple read requests targeting the same object.
+
+  This could be a typical use case where multiple services are setup and teardown very frequently. Image
+  that we have a cluster backing tens of thousands of services. All of them are based on a certain
+  container image. So, those RADOS objects containing the container image would be **very** popular. If
+  a PG is able to serve multiple read requests concurently, these requests are likely to be replied sooner.
+  We will need to drain existing writers before start serving readers, and vice versa.
```

Comment:
Draining is one problem, but it's not everything.  I'd avoid trying to list out the problems and simply state that it should be possible, but would be a substantial effort.